### DISCUSSION TITLE
条件变量的使用

### DISCUSSION THREAD

**Discussion by wangzhankun** [Codesire-Deng/co_context#95] [Q&A]
1. 条件变量在`notify`时，awaiter是在被挂起时的线程调度运行还是在条件变量notify时的当前线程运行？
2. 同一线程是否可以既是条件变量的等待者也是通知者？如下文所示
```cpp

co_context::condition_variable cv;
co_context::mutex cv_m;

std::atomic_bool has_ret{false};

// sendBinary 工作在多个 io_context 上下文中
co_context::task<void>
sendBinary(co_context::socket& sock, const char header[]) {
    DATA data;
    op_t op = {0};

    if (has_ret) {
        has_ret = false;
        auto lk = co_await cv_m.lock_guard();
        cv.notify_all();
    }

    while (async_ecall(ECALL_ENCRYPT_CBC, data.data(), data.size()) < 0) {
        if (get_num_of_ocall_and_eret() > 0) {
            // ecall 失败，容量已满,通知处理返回数据
            auto lk = co_await cv_m.lock_guard();
            cv.notify_all();
        }
        co_await std::suspend_always{};
    }

    {
        auto lk = co_await cv_m.lock_guard();
        co_await cv.wait(cv_m, [] { return get_num_of_ocall_and_eret() > 0; });
        while (!read_ocall_or_eret((uint8_t*)data.data(), &op)) {
            co_await std::suspend_always{};
        }
    }

    co_await (sock.send({header, strlen(header)}, MSG_MORE) && sock.send(data));
}
```

**Comment by Codesire-Deng** [comment]
1. 在哪里挂起就在哪里恢复。原则很简单，一个协程永远在同一个 `io_context` 中运行，除非使用 `lazy::resume_on` 来明示切换。（如果想在 `notify()` 的线程恢复也很好做，只是目前没给 API。）
2. 可以。另外你的代码有比较多 API 是我不了解的，`io_context`不会追踪因`co_await std::suspend_always{};`而暂停的协程，所以这里看起来容易造成协程被永久遗忘（协程泄漏），除非你的调度器已经跟踪了这个协程。  或许可以考虑 `co_await lazy::yield();`？

**Comment by wangzhankun** [comment]
0. `typedef std::array<char, size> DATA;`
1. `async_ecall`的API仅仅做了数据拷贝，拷贝到两个进程共享内存里面
2. `get_num_of_ocall_and_eret` 获取的当前接收到了多少另一个进程的返回值
3. `read_ocall_or_eret` 将另一个进程处理好的数据拷贝到当前协程栈中

`co_await std::suspend_always{};` 的目的就是想在由于共享内存缓冲区已满无法写入的情况下挂起当前协程；或者是当前没有收到另一进程的返回数据时挂起当前协程。
`co_await lazy::yield();`本身是挂起当前协程同时重新调度协程的API吗？

**Reply by Codesire-Deng** [reply]
> co_await lazy::yield();本身是挂起当前协程同时重新调度协程的API吗？

是的，它看上去像 [std::this_thread::yield](https://en.cppreference.com/w/cpp/thread/yield)。协程会被暂停，但又立即排队执行。这保证其它就绪的协程一定有机会执行。
这样看确实 `co_await lazy::yield();` 比 `suspend_always` 更安全一些。

更长远地看，你的调度器最好能支持某种 IO 多路复用机制：当所有远程调用都陷入等待，则线程陷入等待；当至少有一个远程调用有进展，则线程恢复。这样能保证在低负载场景中 CPU 不会空转。

**Comment by wangzhankun** [comment]
在协程挂起时，可以保持锁的锁定状态吗？
比如说：
```cpp
{
    auto lk = co_await cv_m.lock_guard();
    // do something
    if(failed)
    {
        co_await lazy::yield();
    }
}
```

**Reply by Codesire-Deng** [reply]
会保持的。如果需要让出锁，要么使用 `condition_variable`，要么手动 `unlock()`。

**Reply by wangzhankun** [reply]
那在同一个线程上下文中，同一个锁是可以重入的吗？就是在我这个例子中，如果线程A获取到了锁但是挂起了协程，线程A的其他协程可以执行到do something吗？

**Reply by Codesire-Deng** [reply]
同一线程的两个协程可以，但第二个协程会拿不到锁。

同一个协程锁两次就不行，例如连续两行 `co_await mtx.lock()` 会死锁

**Reply by Codesire-Deng** [reply]
这些行为都在模拟线程锁，先忘记`io_context`，类比“协程=线程”一下即可

**Reply by wangzhankun** [reply]
如果一个协程持有了锁，不会阻塞同线程中另一个协程的执行，只会阻塞不同线程的协程的执行，对吗？从锁的持有上，有且仅有一个协程是锁的持有者，换句话说，线程才是锁的持有者。线程之间是对临界区互斥访问的，同一线程中的协程对临界区不是互斥访问的（因为从执行流上讲，同一线程的不同协程蕴含了对临界区的互斥访问）。
如果这样理解是正确的话，那么当使用`lazy::resume_on`切换线程执行时，锁的行为是怎么样的呢？还是说这是一种未定义行为，应当避免？

**Reply by Codesire-Deng** [reply]
> 如果一个协程持有了锁，不会阻塞同线程中另一个协程的执行，只会阻塞不同线程的协程的执行，对吗？

不是的，**协程之间的 mutex 互斥与所在线程无关**。虽然从执行流上看，同一线程的不同协程确实不会同时执行，但这仍不能避免数据竞争。试想同一线程中两个协程欲对同一个 buffer 进行读-改-写操作，第一个协程对 buffer 上锁后，因等待网络而暂停，它是不希望第二个协程拿到这个锁的，因为此时 buffer 的内容正处于修改当中。

所以，`co_context::mutex` 的设计是“最多只被一个协程占有，协程暂停时不会让出锁”。这也是其它语言（例如 Go，C#）的做法。

> 当使用 `lazy::resume_on` 切换线程执行时，锁的行为是怎么样的呢

由于 `co_context::mutex` 的行为与线程无关，`lazy::resume_on` 可以带着锁一起转移。
